src/core/filesystem.cc

```diff
@@ -555,16 +601,21 @@ S3FileSystem::ParsePath(
     const std::string& path, std::string* bucket, std::string* object)
 {
   // Get the bucket name and the object path. Return error if input is malformed
-  int bucket_start = path.find("s3://") + 5;
-  int bucket_end = path.find("/", bucket_start);
-
-  // If there isn't a second slash, the address has only the bucket
-  if (bucket_end > bucket_start) {
-    *bucket = path.substr(bucket_start, bucket_end - bucket_start);
-    *object = path.substr(bucket_end + 1);
-  } else {
-    *bucket = path.substr(bucket_start);
-    *object = "";
+  re2::RE2 s3_regex(
+      "s3://([a-z]+):([0-9]+)/([0-9a-z.-]+)(((/[0-9a-z.-_]+)*)?)");
+  std::string host_name, host_port;
+  if (!RE2::FullMatch(path, s3_regex, &host_name, &host_port, bucket, object)) {
+    int bucket_start = path.find("s3://") + 5;
```

Comment:
+ strlen("s3://")
Documents better what you are doing and compiler will optimize it to constant 5.